src/Tools/dotnet-monitor/Profiler/ProfilerService.cs

```diff
@@ -98,10 +98,31 @@ public async Task ApplyProfiler(IEndpointInfo endpointInfo, CancellationToken ca
                         endpointInfo.RuntimeInstanceCookie.ToString("D"),
                         cancellationToken);
 
-                    await client.SetStartupProfilerAsync(
-                        ProfilerIdentifiers.Clsid.Guid,
-                        profilerFileInfo.PhysicalPath,
-                        cancellationToken);
+                    // There's no way to ask the target process if it is suspended at startup.
+                    // Just attempt to attach the profiler; if the process is already running,
+                    // this will throw an exception. Then fallback to setting the startup profiler,
+                    // assuming that the failure occurred due to the process already running.
+                    // Can't check that assumption because ServerErrorException and related
+                    // exceptions do not report what the issue is beyond the human readable
+                    // Exception.Message property value.
+                    try
+                    {
+                        // This will wait until the profiler's ICorProfilerCallback3::InitializeForAttach
+                        // implementation completes.
+                        await client.AttachProfilerAsync(
+                            TimeSpan.FromSeconds(10),
+                            ProfilerIdentifiers.Clsid.Guid,
+                            profilerFileInfo.PhysicalPath,
+                            Array.Empty<byte>(),
+                            cancellationToken);
+                    }
+                    catch (ServerErrorException)
+                    {
+                        await client.SetStartupProfilerAsync(
```

Comment:
I was really expecting the reverse here; we SetStartupProfiler by default, and if that fails we fall back to Attach. What error codes does SetStartupProfilerAsync return if the process is not suspended?

Reply:
We should be able to do either way; they should be mutually exclusive.

Reply:
I'll flip it around so that SetStartupProfiler is first; assuming processes that are connecting to the .NET Monitor server are more likely to not use nosuspend than are to use nosuspend, then this call should succeed more often than calling AttachProfiler.

Reply:
I investigated this a bit more and we can't flip this around to what you'd expect. If the process has `nosuspend` and is already running managed code, the `SetStartupProfiler` command will succeed; this repros on .NET 6 and .NET 7 (haven't tested .NET 8, but it probably repros there too.

fyi @davmason

Reply:
I've updated the comments to reflect what is happening and why we cannot do SetStartupProfiler first.